vllm/engine/spec_dec.py

```diff
@@ -0,0 +1,138 @@
+from vllm.config import SpecDecConfig
+from vllm.model_executor import get_model
+from vllm.sequence import SequenceGroupMetadata
+from transformers import AutoModelForCausalLM
+import torch
+from typing import List, Dict
+from vllm.sequence import SamplerOutput, SequenceGroupOutputs, SequenceOutputs
+from vllm.worker.worker import Worker
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+# FIXME: we should get pad_token_id from tokenizer
+PAD_TOKEN_ID = 0
+
+class SpecDecWorker(Worker):
+    def __init__(self, config: SpecDecConfig) -> None:
+        self.propose_cnt = config.propose_cnt
+        self.draft_model_config = config.draft_model_config
+        
+        # self.draft_model = get_model(self.draft_model_config)
+        logger.info(
+            "Initializing speculative decoding worker: "
+            f"model={self.draft_model_config.model!r}, "
+            f"tokenizer={self.draft_model_config.tokenizer!r}, "
+            f"propose_cnt={self.propose_cnt}, "
+            f"seed={self.draft_model_config.seed})")
+        self.draft_model = AutoModelForCausalLM.from_pretrained(self.draft_model_config.model).cuda()
+        
+        ##### values to be set
+        self.draft_probs = None
+        self.draft_kvs = None # if we use hf stype kvs
+
+    def _prepare_inputs(self, 
+                        seq_group_metadata_list: List[SequenceGroupMetadata]) -> List[torch.Tensor]:
+        input_ids_list = []
+        for seq_group_metadata in seq_group_metadata_list:
+            assert len(seq_group_metadata.seq_data) == 1, f"Speculative Decoding does nor beam search for now: {len(seq_group_metadata.seq_data)}"
+            seq_id = next(iter(seq_group_metadata.seq_data))
+            seq = seq_group_metadata.seq_data[seq_id]
+            input_ids_list.append(seq.get_token_ids())
+        max_len = max([len(input_ids) for input_ids in input_ids_list])
+        input_ids_list = [_pad_left_to_max(input_ids, max_len, PAD_TOKEN_ID) for input_ids in input_ids_list]
+        return torch.tensor(input_ids_list, dtype=torch.long, device='cuda')
+    
+    # TODO: we need to align draft and target model's sampler
+    def _sample_method(self, logits):
+        temperature = 1.0
+        return torch.softmax(logits / temperature, dim=-1)
+    
+    def set_draft_tokens(self,
+                seq_group_list: List[SequenceGroupMetadata]) -> None:
+        logger.info(f"# of input request: {len(seq_group_list)}")
```

Comment:
You may want to remove this kind of logs? I am wondering if it may affect the performance since it will print multiple times even for one request. Or there is any flag that helps to avoid overhead when it is running in production env?